pkg/services/rpcsrv/server.go

```diff
@@ -2666,6 +2699,58 @@ drainloop:
 	close(s.notaryRequestCh)
 }
 
+// handleSubscriptions is a single point of blockchain subscriptions/unsubscriptions handling,
+// it is needed to prevent ambiguous situations with subscriptions from different threads.
+func (s *Server) handleSubscriptions() {
+subloop:
+	for {
+		select {
+		case sub := <-s.subCh:
+			switch ch := sub.(type) {
+			case chan *block.Block:
+				s.chain.SubscribeForBlocks(ch)
+			case chan *transaction.Transaction:
+				s.chain.SubscribeForTransactions(ch)
+			case chan *state.ContainedNotificationEvent:
+				s.chain.SubscribeForNotifications(ch)
+			case chan *state.AppExecResult:
+				s.chain.SubscribeForExecutions(ch)
+			case chan mempoolevent.Event:
+				s.coreServer.SubscribeForNotaryRequests(ch)
+			default:
+				panic(fmt.Sprintf("bad subscription: %T", sub))
+			}
+		case unsub := <-s.unsubCh:
+			switch ch := unsub.(type) {
+			case chan *block.Block:
+				s.chain.UnsubscribeFromBlocks(ch)
+			case chan *transaction.Transaction:
+				s.chain.UnsubscribeFromTransactions(ch)
+			case chan *state.ContainedNotificationEvent:
+				s.chain.UnsubscribeFromNotifications(ch)
+			case chan *state.AppExecResult:
+				s.chain.UnsubscribeFromExecutions(ch)
+			case chan mempoolevent.Event:
+				s.coreServer.UnsubscribeFromNotaryRequests(ch)
+			default:
+				panic(fmt.Sprintf("bad unsubscription: %T", unsub))
+			}
+		case <-s.shutdown:
```

Comment:
OK, we probably need some special exit condition, because we have to make several subscriptions after server's shutdown:
```
2022-11-17T20:10:56.6942516Z ==================
2022-11-17T20:10:56.6942684Z 2022/11/17 20:01:52 http: panic serving 127.0.0.1:60060: send on closed channel
2022-11-17T20:10:56.6942829Z goroutine 12745 [running]:
2022-11-17T20:10:56.6942938Z net/http.(*conn).serve.func1()
2022-11-17T20:10:56.6943130Z 	/opt/hostedtoolcache/go/1.18.7/x64/src/net/http/server.go:1825 +0x106
2022-11-17T20:10:56.6943234Z panic({0x12ccf20, 0x169aef0})
2022-11-17T20:10:56.6943426Z 	/opt/hostedtoolcache/go/1.18.7/x64/src/runtime/panic.go:844 +0x258
2022-11-17T20:10:56.6943762Z github.com/nspcc-dev/neo-go/pkg/services/rpcsrv.(*Server).unsubscribeFromChannel(0xc007e18e00, 0x1)
2022-11-17T20:10:56.6944038Z 	/home/runner/work/neo-go/neo-go/pkg/services/rpcsrv/server.go:2551 +0x1c5
2022-11-17T20:10:56.6944415Z github.com/nspcc-dev/neo-go/pkg/services/rpcsrv.(*Server).handleWsReads(0xc007e18e00, 0xc007ead1e0, 0xc007fd7620, 0xc0083f3d40)
2022-11-17T20:10:56.6944697Z 	/home/runner/work/neo-go/neo-go/pkg/services/rpcsrv/server.go:616 +0x551
2022-11-17T20:10:56.6945088Z github.com/nspcc-dev/neo-go/pkg/services/rpcsrv.(*Server).handleHTTPRequest(0xc007e18e00, {0x16a6b70, 0xc006994ee0}, 0xc00156a600)
2022-11-17T20:10:56.6945371Z 	/home/runner/work/neo-go/neo-go/pkg/services/rpcsrv/server.go:463 +0x9eb
2022-11-17T20:10:56.6945565Z net/http.HandlerFunc.ServeHTTP(0xc0079b39f0, {0x16a6b70, 0xc006994ee0}, 0x1?)
2022-11-17T20:10:56.6945757Z 	/opt/hostedtoolcache/go/1.18.7/x64/src/net/http/server.go:2084 +0x4e
2022-11-17T20:10:56.6945967Z net/http.serverHandler.ServeHTTP({0xc0083add10?}, {0x16a6b70, 0xc006994ee0}, 0xc00156a600)
2022-11-17T20:10:56.6946159Z 	/opt/hostedtoolcache/go/1.18.7/x64/src/net/http/server.go:2916 +0x897
2022-11-17T20:10:56.6946314Z net/http.(*conn).serve(0xc007989e00, {0x16a74c0, 0xc0083adbc0})
2022-11-17T20:10:56.6946497Z 	/opt/hostedtoolcache/go/1.18.7/x64/src/net/http/server.go:1966 +0xbab
2022-11-17T20:10:56.6946614Z created by net/http.(*Server).Serve
2022-11-17T20:10:56.6946803Z 	/opt/hostedtoolcache/go/1.18.7/x64/src/net/http/server.go:3071 +0x80d
2022-11-17T20:10:56.6947367Z     logger.go:130: 2022-11-17T20:01:53.882Z	INFO	persisted to disk	{"blocks": 0, "keys": 115, "headerHeight": 0, "blockHeight": 0, "took": "155.901µs"}
```